### PR TITLE
namespace: move namespacesInfo out of cache.

### PR DESCRIPTION
**Pull request by disksing**
This PR tries to make table namespace classifier more independent from the original modules.
The main idea is `namespacesInfo` is detail of `tableClassifier`, which does not need to be known by `RaftCluster` or `clusterInfo`.

### CODE REVIEW

File: server/cluster.go [hunk 1]
```diff
@@ -119,11 +121,10 @@ func (c *RaftCluster) start() error {
 	var classifier namespace.Classifier
 	if c.s.cfg.EnableNamespace {
 		log.Infoln("use namespace classifier.")
-		nsInfo := c.cachedCluster.namespacesInfo
-		if err := nsInfo.loadNamespaces(c.s.kv, kvRangeLimit); err != nil {
+		classifier, err = newTableNamespaceClassifier(core.DefaultTableIDDecoder, c.s.kv, c.s.idAlloc)
```
Comment (siddontang) [inline server/cluster.go]: I think we can remove the decoder argument here too. We always use the default decoder for the table classifier.

File: server/cluster.go [hunk 2]
```diff
@@ -119,11 +121,10 @@ func (c *RaftCluster) start() error {
 	var classifier namespace.Classifier
 	if c.s.cfg.EnableNamespace {
```
Comment (siddontang) [inline server/cluster.go]: I prefer using classifier name switch later, not EnableNamespace.

File: server/cluster.go [hunk 3]
```diff
@@ -705,104 +706,96 @@ func (c *RaftCluster) putConfig(meta *metapb.Cluster) error {
 	return c.cachedCluster.putMeta(meta)
 }
 
-// GetNamespaces returns all the namespace in etc.d
-func (c *RaftCluster) GetNamespaces() []*Namespace {
-	return c.cachedCluster.getNamespaces()
+// GetNamespaceClassifier returns current namespace classifier.
+func (c *RaftCluster) GetNamespaceClassifier() namespace.Classifier {
+	c.RLock()
```
Comment (siddontang) [inline server/cluster.go]: why not saving the classifier in the RaftCluster directly, so we can return it directly, no need Lock here.

File: server/cluster.go [hunk 4]
```diff
@@ -705,104 +706,96 @@ func (c *RaftCluster) putConfig(meta *metapb.Cluster) error {
...
+func (c *RaftCluster) GetNamespaces() []*Namespace {
+	classifier := c.GetNamespaceClassifier()
+	if classifier == nil {
+		return nil
 	}
-
-	cachedCluster := c.cachedCluster
-	if n := cachedCluster.getNamespace(name); n != nil {
-		return errors.New("Duplicate namespace Name")
+	tableClassifier, ok := classifier.(*tableNamespaceClassifier)
```
Comment (siddontang) [inline server/cluster.go]: why not using `classifier.GetNamespaces()` directly?

File: server/cluster.go [hunk 5]
```diff
@@ -705,104 +706,96 @@ func (c *RaftCluster) putConfig(meta *metapb.Cluster) error {
...
-	r := regexp.MustCompile(`^\w+$`)
-	matched := r.MatchString(name)
-	if !matched {
-		return errors.New("Name should be 0-9, a-z or A-Z")
+// GetNamespaces returns all the namespace in etcd.
+// FIXME: remove it after refactor namespace API.
+func (c *RaftCluster) GetNamespaces() []*Namespace {
+	classifier := c.GetNamespaceClassifier()
+	if classifier == nil {
+		return nil
```
Comment (siddontang) [inline server/cluster.go]: I think we must have a classifier now.

File: server/cluster.go (hunk 1 above)
Comment (disksing) [reply server/cluster.go]: Agree. But I prefer to keep this PR's intent accurate. We can do it later.

File: server/cluster.go (hunk 2 above)
Comment (disksing) [reply server/cluster.go]: Agree. But I prefer to keep this PR's intent accurate. We can do it later.

File: server/cluster.go (hunk 4 above)
Comment (disksing) [reply server/cluster.go]: `classifier.GetNamespaces()` returns `[]string`, while we need `[]*Namespace` here.

File: server/cluster.go (hunk 3 above)
Comment (disksing) [reply server/cluster.go]: Addressed, now it is created inside server, seems more reasonable.

**Comment by siddontang** [comment]
PTAL @nolouch

**Comment by nolouch** [comment]
LGTM

**Pull request merged**